torch/distributed/fsdp/_init_utils.py

```diff
@@ -131,9 +147,23 @@ def _init_process_group_state(
 
 @no_type_check
 def _init_process_group_state_for_hybrid_shard(
-    state: _FSDPState, process_group
+    state: _FSDPState,
+    process_group: ProcessGroupType,
+    device_mesh: DeviceMesh,
 ) -> _FSDPState:
-    if process_group is None:
+    if device_mesh:
+        if _is_valid_hybrid_shard_device_mesh(device_mesh):
+            state._device_mesh = device_mesh
+            # We currenlty only allow _inter_node_pg to be the outermost dimension, and the
+            # process_group(intra_node) to be the innermost dimension.
+            state._inter_node_pg = device_mesh.get_dim_groups(mesh_dim=0)
+            state.process_group = device_mesh.get_dim_groups(mesh_dim=1)
+        else:
+            raise ValueError(
+                "Expected device_mesh to be passed in has ndim=2 "
+                f"but got {len(device_mesh.get_dim_groups())}"
+            )
```

Comment:
Can you mention that we do not support HSDP + TP at this moment?

Reply:
Let's maybe put this in the 2D tutorial that we are going to make.